### PR TITLE
Change wallet password

### PR DESCRIPTION
**Pull request by Sirmorrison**
### Resolves

Issue #58

### What's new

This PR adds the ability change wallet passphrase.
It also moves the ability to modify wallet information to a dialogue.

### Relevant screenshots or logs
image

image

image

image

image

image

### Additional context

_Add any other context about the problem here._

### CODE REVIEW

**Review by metaclips — CHANGES_REQUESTED** [review]
Use single line editor when changing wallet name so that this wont occur.
Screenshot 2020-04-11 at 15 47 15

When changing wallet name and error occurs, you should show error thrown.

**Comment by Sirmorrison** [comment]
> Use single line editor when changing wallet name so that this wont occur.
> Screenshot 2020-04-11 at 15 47 15
> 
> When changing wallet name and error occurs, you should show error thrown.

Out of scope of work.

**Comment by metaclips** [comment]
> > Use single line editor when changing wallet name so that this wont occur.
> > Screenshot 2020-04-11 at 15 47 15
> > When changing wallet name and error occurs, you should show error thrown.
> 
> Out of scope of work.

True. An issue should be created so we dont miss it later.

File: ui/handle_input.go [hunk 1]
```diff
@@ -236,42 +277,102 @@ func (win *Window) HandleInputs() {
...
+		case strength > 0.6:
+			win.outputs.passwordStgth.ProgressColor = win.theme.Color.Success
+			win.outputs.passStgthTxt.Text = "Strong"
+		case strength > 0.3 && strength <= 0.6:
+			win.outputs.passStgthTxt.Text = "moderate"
+			win.outputs.passwordStgth.ProgressColor = win.theme.Color.Average
+		default:
+			win.outputs.passStgthTxt.Text = "Weak"
+			win.outputs.passwordStgth.ProgressColor = win.theme.Color.Danger
+		}
```
Comment (metaclips) [inline ui/handle_input.go]: Do remove all password strength text as it's no longer needed

File: ui/dialogs.go [hunk 2]
```diff
@@ -142,3 +147,93 @@ func (win *Window) infoDiag() {
...
+							})
+						}),
+					)
+				}),
+			)
+		})
+	})
+}
+
+func (win *Window) passwordStrenght() {
```
Comment (metaclips) [inline ui/dialogs.go]: passwordStrength

File: ui/handle_input.go [hunk 3]
```diff
@@ -126,6 +126,46 @@ func (win *Window) HandleInputs() {
 		}
 	}
 
+	// CHANGE WALLET PASSWORD
+
+	if win.inputs.changePasswordDiag.Clicked(win.gtx) {
```
Comment (metaclips) [inline ui/handle_input.go]:
```suggestion
	for win.inputs.changePasswordDiag.Clicked(win.gtx) {
```
So as to catch subsequent clicks

File: ui/handle_input.go [hunk 4]
```diff
@@ -126,6 +126,46 @@ func (win *Window) HandleInputs() {
...
+		win.dialog = win.editPasswordDiag
+		win.states.dialog = true
+	}
+
+	if win.inputs.savePassword.Clicked(win.gtx) {
```
Comment (metaclips) [inline ui/handle_input.go]: Use a for loop instead

File: ui/handle_input.go [hunk 5]
```diff
@@ -126,6 +126,46 @@ func (win *Window) HandleInputs() {
...
+		}
+		np := win.validatePasswords()
+		if np == "" {
+			return
+		}
+
+		err := win.wallet.ChangeWalletPassphrase(win.walletInfo.Wallets[win.selected].ID, op, np)
+		if err != nil {
+			log.Debug("Error changing wallet password " + err.Error())
+			if err.Error() == "invalid_passphrase" {
```
Comment (metaclips) [inline ui/handle_input.go]:
```suggestion
			if err.Error() == dcrlibwallet.ErrInvalidPassphrase {
```

**Review by metaclips — COMMENTED** [review]
UI functionality looks good by the way.

**Review by metaclips — DISMISSED** [review]

File: ui/decredmaterial/progressbar.go [hunk 6]
```diff
@@ -15,6 +15,7 @@ import (
 type ProgressBar struct {
 	BackgroundColor color.RGBA
 	ProgressColor   color.RGBA
+	Progress        float64
 }
```
Comment (oshorefueled) [inline ui/decredmaterial/progressbar.go:19]: any reason for making modifications to the progressbar code?

**Review by oshorefueled — CHANGES_REQUESTED** [review]

File: ui/widgets.go [hunk 7]
```diff
@@ -46,20 +46,20 @@ type outputs struct {
...
+	spendingPassword, matchSpending, oldSpendingPassword, dialog, rename       decredmaterial.Editor
 	toOverview, toWallets, toTransactions, toRestoreWallet, toSend, toSettings decredmaterial.IconButton
 	toReceive                                                                  decredmaterial.IconButton
 	createDiag, cancelDiag, addAcctDiag                                        decredmaterial.IconButton
 
 	createWallet, restoreDiag, restoreWallet, deleteWallet, deleteDiag, gotItDiag decredmaterial.Button
 	toggleWalletRename, renameWallet, syncHeader                                  decredmaterial.IconButton
-	sync, moreDiag                                                                decredmaterial.Button
+	sync, moreDiag, changePasswordDiag, savePassword                              decredmaterial.Button
+	passwordStgth                                                                 *decredmaterial.ProgressBar
```
Comment (oshorefueled) [inline ui/widgets.go]: kindly rename "passwordStgth" this

File: ui/handle_input.go [hunk 8]
```diff
@@ -236,42 +276,99 @@ func (win *Window) HandleInputs() {
...
 		clipboard.WriteAll(win.walletInfo.Wallets[win.selected].Accounts[win.selectedAccount].CurrentAddress)
 		win.addressCopiedLabel.Text = "Address Copied"
 		time.AfterFunc(time.Second*3, func() {
 			win.addressCopiedLabel.Text = ""
 		})
 		return
 	}
+
+	if strings.Trim(win.inputs.spendingPassword.Text(), " ") != "" {
+		strength := (dcrlibwallet.ShannonEntropy(win.inputs.spendingPassword.Text()) / 4.0)
```
Comment (oshorefueled) [inline ui/handle_input.go]: parentheses around "dcrlibwallet.ShannonEntropy(win.inputs.spendingPassword.Text()) / 4.0" is redundant.

**Review by oshorefueled — DISMISSED** [review]

File: ui/decredmaterial/theme.go [hunk 9]
```diff
@@ -36,11 +36,13 @@ type Theme struct {
 		Secondary  color.RGBA
 		Text       color.RGBA
 		Hint       color.RGBA
+		Overlay    color.RGBA
 		InvText    color.RGBA
 		Success    color.RGBA
 		Danger     color.RGBA
 		Background color.RGBA
 		Surface    color.RGBA
+		Average    color.RGBA
```
Comment (oluwandabira) [inline ui/decredmaterial/theme.go]: Should be a page color

**Review by oluwandabira — CHANGES_REQUESTED** [review]

**Review by oshorefueled — DISMISSED** [review]

File: ui/handle_input.go [hunk 10]
```diff
@@ -248,42 +289,99 @@ func (win *Window) HandleInputs() {
...
+
+		win.outputs.passwordBar.Progress = strength * 100
+	}
+
+	if win.err != "" {
+		time.AfterFunc(time.Second*3, func() {
+			win.err = ""
+			win.outputs.err.Color = win.theme.Color.Danger
+		})
+		return
```
Comment (oluwandabira) [inline ui/handle_input.go]: This is unsafe. `AfterFunc` calls the func on its own go routine

**Review by oluwandabira — CHANGES_REQUESTED** [review]

File: ui/handle_input.go [hunk 11]
```diff
@@ -248,42 +307,91 @@ func (win *Window) HandleInputs() {
...
 	if match != pass {
-		win.outputs.matchSpending.Color = win.theme.Color.Danger
+		win.err = "New wallet passwords does no match. Try again."
 		return ""
 	}
 
 	return pass
 }
 
+func (win *Window) valOldAndNewPasswords() string {
```
Comment (oluwandabira) [inline ui/handle_input.go]: I don't see why this is necessary. Password requirements are for dcrlibwallet to handle.

**Review by oluwandabira — CHANGES_REQUESTED** [review]

File: ui/handle_input.go [hunk 12]
```diff
@@ -281,9 +353,23 @@ func (win *Window) validatePasswords() string {
...
+
+	return pass
+}
+
+func (win *Window) validateOldPassword() string {
+	pass := win.inputs.oldSpendingPassword.Text()
+	if strings.Trim(pass, " ") == "" {
+		win.outputs.oldSpendingPassword.HintColor = win.theme.Color.Danger
+		win.err = "Old Wallet password required and cannot be empty"
+		return ""
```
Comment (oluwandabira) [inline ui/handle_input.go]: Let `dcrlibwallet` handle empty passwords

**Review by oluwandabira — CHANGES_REQUESTED** [review]

**Review by oluwandabira — APPROVED** [review]

**Pull request merged**